### PR TITLE
Job Improvement Bundle patch

### LINKED ISSUE #5430: Changes the Adoramus skill property to Neutral if user consumer Ancilla.
rAthena Hash: https://github.com/rathena/rathena
Client Date: 2020-04-01
Server Mode: Renewal
Description of Issue:
If user consumes Ancilla, changes the "Adoramus" skill property to neutral for the Ancilla duration.

Result:
Official Information:
https://ro.gnjoy.com/guide/runemidgarts/skillview.asp?lineseq=6&skillid=2040

### LINKED ISSUE #5450: Severe Rainstorm skill damage explosion when equip instruments/whips
* **rAthena Hash**: 8385162fd

* **Client Date**: 20200401

* **Server Mode**: RE

* **Description of Issue**: Use Improve Concentration skill right after use Severe Rainstorm skill, the severe rainstorm damage will explosion if you use element arrow or the same element arrow as the monster.

skillbug

  * Result: 
  * Expected Result: the damage should remain the same if you cast improve concentration after you cast severe rainstorm skill.

skillbugNor

  * How to Reproduce: equip arrow of shadow (item#1767), then `@monster 1719`. use improve concentration after you cast severe rainstorm skill, the damage increases a lot. I tried different element arrow, they all affect. You can see in the video below. If you equip steel arrow(item#1753), then the damage does not affect by improve concentration skill.
  * Official Information:
  

* **Modifications that may affect results**: no mod. I use clean Ra to test.
  
  

Video: https://www.youtube.com/watch?v=SN5vf8yakfY

### PR DESCRIPTION
**Pull request by aleos89**
* **Addressed Issue(s)**: Fixes #5430 and fixes #5450.

* **Server Mode**: Renewal

* **Description of Pull Request**: 
  * Updates various skills following kRO's job improvement patch.
  * Change log: https://ro.gnjoy.com/news/notice/View.asp?BBSMode=10001&seq=7275&curpage=2
Thanks to @ecdarreola!

TODO Homunculus changes:
- The experience required on level 1~99 is greatly reduced.
- Increases bonus HP and SP gained from raising base level of basic homunculus.
- Increases bonus HP and SP gained from evolving homunculus.
- Increases bonus HP and SP gained from raising base level of homunculus S.
- Increases bonus HP and SP gained from changing homunculus to homunculus S.
(The HP and SP increasing effects only apply to new homunculi that are created after this update).

### CODE REVIEW

File: src/map/battle.cpp [hunk 1]
```diff
@@ -2617,7 +2632,7 @@ static bool is_attack_critical(struct Damage* wd, struct block_list *src, struct
 			case SN_SHARPSHOOTING:
 			case MA_SHARPSHOOTING:
 #ifdef RENEWAL
-				cri += 300; // !TODO: Confirm new bonus
+				cri += cri / 2;
```
Comment (attackjom) [inline src/map/battle.cpp:2635]: It said Critical modifier from itembonus half not critical rate.

File: src/map/battle.cpp [hunk 2]
```diff
@@ -7330,42 +7325,33 @@ int64 battle_calc_return_damage(struct block_list* bl, struct block_list *src, i
...
+					status_change_end(bl, SC_DEATHBOUND, INVALID_TIMER);
+					rdamage += rd1 * 70 / 100; // Target receives 70% of the amplified damage. [Rytech]
 				}
+			}
 
-				if( sc->data[SC_SHIELDSPELL_DEF] && sc->data[SC_SHIELDSPELL_DEF]->val1 == 2 && !status_bl_has_mode(src,MD_STATUS_IMMUNE) ){
-						rdamage += damage * sc->data[SC_SHIELDSPELL_DEF]->val2 / 100;
-						if (rdamage < 1) rdamage = 1;
-				}
+			if( sc->data[SC_SHIELDSPELL_DEF] && sc->data[SC_SHIELDSPELL_DEF]->val1 == 2 && !status_bl_has_mode(src,MD_STATUS_IMMUNE) ){
```
Comment (attackjom) [inline src/map/battle.cpp:7352]: Shield spell is totally rework to regen SP/HP , increase ATK/MATK.

File: src/map/skill.cpp [hunk 3]
```diff
@@ -5939,6 +5926,17 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
 			skill_attack(BF_MAGIC,src,src,bl,skill_id,skill_lv,tick,flag);
 		break;
 
+	case LG_OVERBRAND: {
```
Comment (attackjom) [inline src/map/skill.cpp:5929]: from this video https://www.youtube.com/watch?v=0gI7Cwl8yvk

Overbrand has simple Circular AOE around caster.

**Comment by tumemmac** [comment]
Job Rebellion:

**Dragon tail** - The skill logic has been changed to cast directly on the target target,
and if it is in the "Crimson Marker" state , it will be changed to deal higher damage to both the target and the target within the surrounding range.

source: https://ro.gnjoy.com/news/notice/View.asp?BBSMode=10001&seq=7081

File: src/map/battle.cpp (hunk 1 above)
Comment (aleos89) [reply src/map/battle.cpp:2635]: Sigma has it quoted as `-The half of user's Cri chance is added to the skill base cri chance.`. Did I interpret this wrong?

File: src/map/battle.cpp (hunk 2 above)
Comment (aleos89) [reply src/map/battle.cpp:7352]: Yeah sorry, this was also on the TODO list. I had skipped it and forgot to go back to it.

File: src/map/battle.cpp (hunk 1 above)
Comment (rye305) [reply src/map/battle.cpp:2635]:
```
#ifdef RENEWAL
	if (is_attack_critical(&wd, src, target, skill_id, skill_lv, false)) {
		if (sd) { //Check for player so we don't crash out, monsters don't have bonus crit rates [helvetica]
			wd.damage = (int64)floor((float)((wd.damage * 140) / 100 * (100 + sd->bonus.crit_atk_rate/2)) / 100); //Crit Stuff Here
			if (is_attack_left_handed(src, skill_id))
				wd.damage2 = (int64)floor((float)((wd.damage2 * 140) / 100 * (100 + sd->bonus.crit_atk_rate/2)) / 100); //Crit Stuff Here
		} else
			wd.damage = (int64)floor((float)(wd.damage * 140) / 100);

		if (tsd && tsd->bonus.crit_def_rate != 0) {
			ATK_ADDRATE(wd.damage, wd.damage2, -tsd->bonus.crit_def_rate);
		}
	}
```

**Comment by aleos89** [comment]
@tumemmac That's an old patch note for extended classes. Rebellion was done in #4428.

**Comment by Haydrich** [comment]
Royal Guard
1- there is something strange about ray of genesis 
without inspiration
image
when inspiration were activated
image
the damage here is too high 

2- about Overbrand already not making any damage 
image

3- moonslasher not giving the bonus icon - Overbrand Ready

4- Reflect Damage Reduction
there is something, skill reducing the reflect damage yeah but from level 4 it looks like 100%
but the new formula
```reflect damage taken to reduce damage taken from melee reflect damage by (10 x skill level)%.```

5- Hesperus Lit
without Inspiration 
image
when inspiration were activated
image

**Comment by Haydrich** [comment]
4.5. Blood Sucker
- Reworks skill, consumes 5 ea of Blood Sucker Plant Seed to **give a buff to 1 party member**, target gains ((-1) + (2 x skill level))% chance to restore (skill level)% of damage as HP when dealing physical damage. Skill duration : ((-20) + (60 x skill level)) seconds.
now its targeting normal players

File: src/map/status.cpp [hunk 4]
```diff
@@ -11511,6 +11507,19 @@ int status_change_start(struct block_list* src, struct block_list* bl,enum sc_ty
 			val2 = (status->int_ + status->luk) * val1 / 20 * status_get_lv(bl) / 200 + val1;	// Chance to evade magic damage.
 			val3 = ((val1 * 15) + (10 * (sd?pc_checkskill(sd,CR_DEFENDER):skill_get_max(CR_DEFENDER)))) * status_get_lv(bl) / 100; // Defence added
 			break;
+		case SC_SHIELDSPELL_HP:
+			val2 = 3; // 3% HP every 3 seconds
```
Comment (Haydrich) [inline src/map/status.cpp:11515]: related to Sigma's info its 5% for each 3 seconds
```2.2. Shield Spell
- Level-specific effect will be unified into only 1 effect, changes the skill effect as below.
Level 1 : Recovers 5% of user's MaxHP every 3 seconds for 90 seconds.
Level 2 : Recovers 3% of user's MaxSP every 5 seconds for 90 seconds.
Level 3 : Atk + 150 and Matk + 150 for 90 seconds.```

Comment (aleos89) [reply src/map/status.cpp:11515]: I took that from the kRO skill description. Not sure if Sigma or Gravity made the typo.

Comment (Haydrich) [reply src/map/status.cpp:11515]: yeah, im sry depending on kRO info its 5%

**Comment by Haydrich** [comment]
Ray of Genesis damage still OP, is that right?
image
image

**Comment by Haydrich** [comment]
this icons not appearing : needle of paralyze, call homunculus, overbrand ready and spore explosion

8.6. Killing Cloud
- Removes Red Gemstone consumption, the skill duration is 5 seconds. Damage formula : (40 x skill level)%Matk.
- Replaces poison inflicting effect with new effect "Cloud Poison". Target will be inflicted by Cloud Poison for 5 seconds.
**-Cloud Poison : target will take more damage from poison property by (5 x skill level)%.**
Cloud Poison effect not working, i tried to damage the target with poison element and the damage not increased

**Comment by Haydrich** [comment]
Tiger Cannon: damage improvement that gains from the combo -Dragon Combo and Fallen Empire- not working

**Comment by Akaineko-astasi** [comment]
Tiger Cannon is now a self skill. The zone is the same.

**Comment by aleos89** [comment]
Tiger Cannon works just fine for me. Using it as itself I get about 9,500 damage. Doing Dragon Combo -> Fallen Empire -> Tiger Cannon gets me about 15,000 damage.

**Comment by Haydrich** [comment]
> Tiger Cannon works just fine for me. Using it as itself I get about 9,500 damage. Doing Dragon Combo -> Fallen Empire -> Tiger Cannon gets me about 15,000 damage.

you are right, but there is an issue, when you target more than 1monster for ex : @monster 2413 2
the amplified damage gains to 1 of the 2 monster 
image

**Comment by Haydrich** [comment]
Flash Combo not damage to the nearby monsters, just damage the target only, 
image
from this GiF flash combo damage all monsters around " i mean tiger cannon"

**Comment by Haydrich** [comment]
Poison Cloud effect still not working fine

**Comment by aleos89** [comment]
Can you explain what's wrong with Poison Mist?

**Comment by Haydrich** [comment]
> Can you explain what's wrong with Poison Mist?

using enchant poison and bash, then usinf the skill on player then attack the player with bash, damage should be increased 25% but it still have the same damage

**Comment by Haydrich** [comment]
Now when i use Flash Combo in 2 monsters, Dragon Combo and Fallen Empire target the monster that i target, but Tiger Cannon just target the second monster, and not dealing damage on the target monster

bandicam-2020-11-09-23-24-21-660

**Comment by laziem** [comment]
@Haydrich can you try Pyroclastic? Did the atk inceased shown right away on ALT + A or still need to change map.

**Comment by Haydrich** [comment]
@laziem still need to change your map to get the ATK bonus

**Comment by sonniez** [comment]
Royal Guard
- Genesis ray is not release at max level of Genesis ray user learned. It will random skill level instead when use bAutospell bonus

File: src/map/battle.cpp (hunk 1 above)
Comment (eppc0330) [reply src/map/battle.cpp:2630]: Crit attack rate should be halven on only skill damage, not on normal attack damage. How about this? I have tested it and I think it worked.

> if (is_attack_critical(wd, src, target, skill_id, skill_lv, false)) {
		if (sd) { //Check for player so we don't crash out, monsters don't have bonus crit rates [helvetica]
			if(skill_id > 0)
				wd.damage = (int)floor((float)((wd.damage * 140) / 100 * (100 + (sd->bonus.crit_atk_rate/2))) / 100);
			else
				wd.damage = (int)floor((float)((wd.damage * 140) / 100 * (100 + sd->bonus.crit_atk_rate)) / 100);
			if (is_attack_left_handed(src, skill_id))
				if(skill_id > 0)
					wd.damage2 = (int)floor((float)((wd.damage2 * 140) / 100 * (100 + (sd->bonus.crit_atk_rate/2))) / 100);
				else
					wd.damage2 = (int)floor((float)((wd.damage2 * 140) / 100 * (100 + sd->bonus.crit_atk_rate)) / 100);
		} else
			wd.damage = (int)floor((float)(wd.damage * 140) / 100);

**Comment by Masao87** [comment]
How's progress going on this? Been a little while since the last update :)

**Comment by aleos89** [comment]
Let me know if there are any other issues. If not, I'm not sure how to effectively get the homunculus changes and may be just left on the TODO backburner until a reputable leak is obtained.

**Comment by Razgrom** [comment]
> 
> 
> Let me know if there are any other issues. If not, I'm not sure how to effectively get the homunculus changes and may be just left on the TODO backburner until a reputable leak is obtained.

Hello Aleos! I found a minor bug on this bundle, this skill https://ro.gnjoy.com/guide/runemidgarts/skillview.asp?lineseq=7&skillid=2286  doesn't give Matk during the action of the skill.

**Comment by aleos89** [comment]
@Razgrom If you have any info on whether Auto Shadow Spell should be an equip MATK or normal MATK then I can make the adjustments. It's not showing up in the status window because it's currently flagged as equip MATK. You should see the benefit though in the damage output.

**Comment by Razgrom** [comment]
> 
> 
> @Razgrom If you have any info on whether Auto Shadow Spell should be an equip MATK or normal MATK then I can make the adjustments. It's not showing up in the status window because it's currently flagged as equip MATK. You should see the benefit though in the damage output.

https://www.youtube.com/watch?v=WhqGLcuOvkA&t=123s   28:46

**Comment by eppc0330** [comment]
AM_CALLHOMUN skill still requires item 'Embrio' for creating homunculus. 'Seed of life' item will be required only when recall rest status homunculus.

**Comment by Badarosk0** [comment]
5.3. Rolling Cutter
- Increase the spin counter duration from 5 seconds to 10 seconds.

It is currently only 5 seconds.

**Comment by aleos89** [comment]
@Badarosk0 From my data it has 10 seconds for Rolling Cutter in the skill database. Are you using 3rd class skills in pre-re? I didn't include any of the balancing stuff in the pre-re database as I plan on removing 3rd class skills from the pre-re databases soon.

**Comment by Badarosk0** [comment]
> @Badarosk0 From my data it has 10 seconds for Rolling Cutter in the skill database. Are you using 3rd class skills in pre-re? I didn't include any of the balancing stuff in the pre-re database as I plan on removing 3rd class skills from the pre-re databases soon.

I just downloaded the clean rA and applied this PR. I will check again.

**Comment by aleos89** [comment]
@Razgrom Where did you get that formula for Triangle Shot? There are several instances where I'm not sure how the stat plays a role in different skill damage bonuses.

**Comment by eppc0330** [comment]
Matk bonus of SC_AUTOSHADOWSPELL should be based on level of SC_AUTOSHADOWSPELL, not on level of auto-casting skill.

These are suggestions :

skill.cpp 
```
sc_start4(&sd->bl,&sd->bl,SC__AUTOSHADOWSPELL,100,id,lv,prob,0,skill_get_time(SC_AUTOSHADOWSPELL,aslvl));
```
change **prob,0,skill_get_time** to **prob,aslvl,skill_get_time**
```
sc_start4(&sd->bl,&sd->bl,SC__AUTOSHADOWSPELL,100,id,lv,prob,aslvl,skill_get_time(SC_AUTOSHADOWSPELL,aslvl));
```

status.cpp 
```
matk += sc->data[SC__AUTOSHADOWSPELL]->val2 * 5;
```
Change **val2** to **val4**
```
matk += sc->data[SC__AUTOSHADOWSPELL]->val4 * 5;
```

**Comment by nubspixel** [comment]
https://github.com/rathena/rathena/blob/40fa45630167469bfef56bbc58c123ef2385c003/src/map/battle.cpp#L4646
```
case GN_CARTCANNON:
			// ATK [(20 * skill_lv) + {( INT / (3 - ( Cart Remodeling Skill Level ) )} + ( Cart Cannon Skill Level x 250 )] %
			skillratio += -100 + 250 * skill_lv + 20 * skill_lv + sstatus->int_ / (pc_checkskill(sd, GN_REMODELING_CART) == 3 ? 1 : 3 - pc_checkskill(sd, GN_REMODELING_CART));
			RE_LVL_DMOD(100);
			break;
```
this line cause crash because division by 0.
when you level GN_REMODELING_CART = 3 (minimum level), and fire cannon = crash.

**Comment by eppc0330** [comment]
> https://github.com/rathena/rathena/blob/40fa45630167469bfef56bbc58c123ef2385c003/src/map/battle.cpp#L4646
> 
Current formula of skill Cart cannon is :
250 × cart cannon skilllv + 20 × cart cannon skilllv × cart remodeling skilllv + 2 × INT ÷ (6 - cart remodeling skilllv)

Source will be :
```
-100 + 250 * skill_lv + 20 * skill_lv * pc(checkskill(sd, GN_REMODELING_CART) + 2 * sstatus->int_ / (6 - pc(checkskill(sd, GN_REMODELING_CART));
```

**Comment by eppc0330** [comment]
> @Razgrom Where did you get that formula for Triangle Shot? There are several instances where I'm not sure how the stat plays a role in different skill damage bonuses.

The patch note link above says 
```
For 10Lv Triangle Shot & 200 AGI : Skill ratio will be 2900%
```
So AGI formula will be AGI×3.

Also source will be :
```
-100 + 230 * skill_lv + 3 * sstatus->agi;
```

**Comment by eppc0330** [comment]
AM_CALLHOMUN not works well yet. Can you check source again? REST STATUS homunculus is still called successfully regardless a character has item 'Seed of life' or not.

**Comment by Badarosk0** [comment]
Skill SO_PSYCHIC_WAVE seems to be generating a crash map server. I will still find out how it happens.
```
Program received signal SIGSEGV, Segmentation fault.
0x000000000046232b in battle_calc_magic_attack (src=src@entry=0x7fffd52da4c4, 
    target=target@entry=0x103ca270, skill_id=skill_id@entry=2449, 
    skill_lv=skill_lv@entry=5, mflag=mflag@entry=0) at battle.cpp:6571
6571		switch(skill_id) {
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.212.el6_10.3.x86_64 keyutils-libs-1.4-5.el6.x86_64 krb5-libs-1.10.3-65.el6.x86_64 libcom_err-1.41.12-24.el6.x86_64 libgcc-4.4.7-23.el6.x86_64 libselinux-2.0.94-7.el6.x86_64 libstdc++-4.4.7-23.el6.x86_64 openssl-1.0.1e-58.el6_10.x86_64 pcre-7.8-7.el6.x86_64 zlib-1.2.3-29.el6.x86_64
(gdb) 
(gdb) bt full
#0  0x000000000046232b in battle_calc_magic_attack (
    src=src@entry=0x7fffd52da4c4, target=target@entry=0x103ca270, 
    skill_id=skill_id@entry=2449, skill_lv=skill_lv@entry=5, 
    mflag=mflag@entry=0) at battle.cpp:6571
        i = <optimized out>
        skill_damage = 0
        s_ele = <optimized out>
        sd = 0x0
        tsd = 0x103ca270
        sc = 0x7fffd52dab2c
        tsc = 0x103ca970
        ad = {statusAtk = 0, statusAtk2 = 0, weaponAtk = 0, weaponAtk2 = 0, 
          equipAtk = 0, equipAtk2 = 0, masteryAtk = 0, masteryAtk2 = 0, 
          damage = 1, damage2 = 0, type = <incomplete type>, div_ = 1, 
          amotion = 0, dmotion = 400, blewcount = 0, flag = 258, miscflag = 0, 
          dmg_lv = ATK_DEF, isspdamage = false}
        sstatus = <optimized out>
        tstatus = <optimized out>
        flag = {imdef = 0, infdef = 0}
        __FUNCTION__ = "battle_calc_magic_attack"
        skill = <optimized out>
        nk = std::bitset
        mapdata = <optimized out>
#1  0x00000000004665f5 in battle_calc_attack (attack_type=attack_type@entry=2, 
    bl=bl@entry=0x7fffd52da4c4, target=target@entry=0x103ca270, 
    skill_id=skill_id@entry=2449, skill_lv=skill_lv@entry=5, flag=0)
    at battle.cpp:7918
        d = {statusAtk = 0, statusAtk2 = 0, weaponAtk = 0, weaponAtk2 = 0, 
          equipAtk = 0, equipAtk2 = 0, masteryAtk = 0, masteryAtk2 = 0, 
          damage = 1, damage2 = 0, type = <incomplete type>, div_ = 1, 
          amotion = 0, dmotion = 400, blewcount = 0, flag = 258, miscflag = 0, 
          dmg_lv = ATK_DEF, isspdamage = false}
        sd = <optimized out>
#2  0x000000000064746f in skill_attack (attack_type=2, 
    src=src@entry=0x7fffd52da4c4, dsrc=dsrc@entry=0x7fffd9a5c24c, 
    bl=bl@entry=0x103ca270, skill_id=skill_id@entry=2449, 
    skill_lv=skill_lv@entry=5, tick=29237595666, flag=0) at skill.cpp:3426
        dmg = {statusAtk = 0, statusAtk2 = 0, weaponAtk = 0, weaponAtk2 = 0, 
          equipAtk = 0, equipAtk2 = 0, masteryAtk = 0, masteryAtk2 = 0, 
          damage = 1, damage2 = 0, type = <incomplete type>, div_ = 1, 
          amotion = 0, dmotion = 400, blewcount = 0, flag = 258, miscflag = 0, 
          dmg_lv = ATK_DEF, isspdamage = false}
        sstatus = 0x7fffd52daa98
        tstatus = 0x103ca8e8
        sc = 0x7fffd52dab2c
        tsc = <optimized out>
        sd = 0x0
        tsd = 0x103ca270
        damage = <optimized out>
        rmdamage = false
        type = <optimized out>
        dmg_type = <optimized out>
        shadow_flag = false
        additional_effects = true
        __FUNCTION__ = "skill_attack"
#3  0x000000000065554f in skill_unit_onplace_timer (unit=0x7fffd9a5c24c, 
    bl=0x103ca270, tick=29237595666) at skill.cpp:15158
        sc = <optimized out>
        tsc = 0x103ca970
        skill_id = 2449
        inf2 = <optimized out>
        sg = 0x7ffff11da794
        ss = 0x7fffd52da4c4
        tsd = <optimized out>
        tstatus = 0x103ca8e8
        type = <incomplete type>
        ts = <optimized out>
        tick = 29237595666
        unit = 0x7fffd9a5c24c
        bl = 0x103ca270
#4  0x00000000006575ae in skill_unit_timer_sub_onplace (bl=0x103ca270, 
    ap=<optimized out>) at skill.cpp:20173
        unit = 0x7fffd9a5c24c
        group = 0x7ffff11da794
        tick = 29237595666
        __FUNCTION__ = "skill_unit_timer_sub_onplace"
        skill = std::shared_ptr (count 4, weak 0) <optimized out>
#5  0x000000000050230c in map_foreachinrangeV (
    func=0x6573d0 <skill_unit_timer_sub_onplace(block_list*, __va_list_tag*)>, 
    center=0x7fffd9a5c24c, range=<optimized out>, type=<optimized out>, 
    ap=ap@entry=0x7fffffffd9b8, wall_check=true) at map.cpp:730
        bx = <optimized out>
        by = <optimized out>
        m = 935
        returnCount = 0
        bl = <optimized out>
        blockcount = <optimized out>
        i = 0
        x0 = <optimized out>
        x1 = <optimized out>
        y0 = <optimized out>
        y1 = 93
        ap_copy = {{gp_offset = 48, fp_offset = 48, 
            overflow_arg_area = 0x7fffffffda90, 
            reg_save_area = 0x7fffffffd9d0}}
#6  0x000000000050252f in map_foreachinrange (
    func=func@entry=0x6573d0 <skill_unit_timer_sub_onplace(block_list*, __va_list_tag*)>, center=center@entry=0x7fffd9a5c24c, range=<optimized out>, 
    type=<optimized out>) at map.cpp:746
        returnCount = 0
        ap = {{gp_offset = 32, fp_offset = 48, 
            overflow_arg_area = 0x7fffffffda90, 
            reg_save_area = 0x7fffffffd9d0}}
#7  0x00000000006140cd in skill_unit_timer_sub(DBKey, DBData *, typedef __va_list_tag __va_list_tag *) (key=..., data=<optimized out>, ap=<optimized out>)
    at skill.cpp:20426
        unit = 0x7fffd9a5c24c
        group = 0x7ffff11da794
        tick = 29237595666
        dissonance = false
        bl = 0x7fffd9a5c24c
        __FUNCTION__ = "skill_unit_timer_sub"
#8  0x00000000006c885c in db_obj_vforeach (self=0x26c2e48, 
    func=0x613f60 <skill_unit_timer_sub(DBKey, DBData *, typedef __va_list_tag __va_list_tag *)>, args=0x7fffffffdb90) at db.cpp:2040
        argscopy = {{gp_offset = 24, fp_offset = 48, 
            overflow_arg_area = 0x7fffffffdc70, 
            reg_save_area = 0x7fffffffdbb0}}
        db = 0x26c2e48
        sum = 0
        node = 0x896cb68
        parent = <optimized out>
#9  0x00000000006c3c69 in db_obj_foreach (self=<optimized out>, 
    func=<optimized out>) at db.cpp:2086
        args = {{gp_offset = 16, fp_offset = 48, 
            overflow_arg_area = 0x7fffffffdc70, 
            reg_save_area = 0x7fffffffdbb0}}
#10 0x00000000005fdc00 in skill_unit_timer (tid=<optimized out>, 
    tick=29237595666, id=<optimized out>, data=<optimized out>)
    at skill.cpp:20457
No locals.
#11 0x000000000040e02a in do_timer (tick=<optimized out>) at timer.cpp:374
        tid = 10
        diff = 0
#12 main (argc=<optimized out>, argv=<optimized out>) at core.cpp:381
No locals.
(gdb) 
```

**Comment by aleos89** [comment]
I missed some parenthesis for the weapon check. It should look like this:
```cpp
		case SO_PSYCHIC_WAVE:
 			if (sd && (sd->weapontype1 == W_STAFF || sd->weapontype1 == W_2HSTAFF || sd->weapontype1 == W_BOOK))
 				ad.div_ = 2;
 ```

**Comment by Badarosk0** [comment]
> I missed some parenthesis for the weapon check. It should look like this:
> 
> ```c++
> 		case SO_PSYCHIC_WAVE:
>  			if (sd && (sd->weapontype1 == W_STAFF || sd->weapontype1 == W_2HSTAFF || sd->weapontype1 == W_BOOK))
>  				ad.div_ = 2;
> ```

Work now! Thanks

**Comment by eppc0330** [comment]
Very quick update! Thanks a lot first.

About Call Homunculus, AM_CALLHOMUN :

1. Deleting homunculus does not turn off SC_HOMUN_TIME.

2. Sometimes after using AM_REST, AM_CALLHOMUN does not turn on SC_HOMUN_TIME.

3. Nice! now AM_CALLHOMUN consumes Seed_Of_Life for calling REST STATUS homunculus.
However, AM_CALLHOMUN now requires both Germination_Breed and Seed_Of_Life when creating homunculus.


Sorry for commenting again and again with same issue. AM_CALLHOMUN still has problems above.

**Comment by eppc0330** [comment]
AM_CALLHOMUN :

1. Creating homunculus again, right after deleting homunculus, does not turn on SC_HOMUN_TIME.

2. Now AM_CALLHOMUN requires item randomly. Sometimes it consumes correctly and sometimes it consumes both Germination_Breed and Seed_Of_Life, for both creating and recalling REST STATUS homunculus.

**Comment by eppc0330** [comment]
Thank you for suffering.

Still has some bugs.

1. After deleting homunculus, for once AM_CALLHOMUN creates same homunculus as deleted one. In This case, the skill consumes both Germination_Breed and Seed_Of_Life. 

2. After case 1 and using AM_REST, AM_CALLHOMUN  consumes  Germination_Breed, instead of Seed_Of_Life. Else recalling consumes both items.

**Comment by eppc0330** [comment]
YEAH now I think AM_CALLHOMUN works very well..!! I really appreciate for it.

**Comment by Badarosk0** [comment]
I believe something is missing from the HP and SP formula of homunculos. Perhaps the bonus is not being applied.
PR5TY2DWODAYH5JQTF35
Capture

**Comment by aleos89** [comment]
I didn’t implement those homunculus parts, as stated in the todo. I’m not sure if these are just bonuses changed during level up or if the databases got updated as well. Either way, there’s just no easy way to get this data without a leak.

File: db/re/produce_db.txt [hunk 5]
```diff
@@ -387,308 +389,308 @@
...
+278,531,30,0,0,11058,0,512,1,519,1
 // Carrot Juice (534) <-- 1 Carrot, 1 Milk
-278,534,30,0,0,11058,0,515,1,519,1
+279,534,30,0,0,11058,0,515,1,519,1
 // Grape Juice (533) <-- 1 Grape, 1 Milk
-279,533,30,0,0,11058,0,514,1,519,1
+281,533,30,0,0,11058,0,514,1,519,1
 // Unripe Apple (619) <-- 10 Sticky Mucus, 20 Green Herbs, 10 Apples
-280,619,30,0,0,11058,0,938,10,511,20,512,10
+281,619,30,0,0,11058,0,938,10,511,20,512,10
```
Comment (Badarosk0) [inline db/re/produce_db.txt:696]:
```
282,619,30,0,0,11058,0,938,10,511,20,512,10
```

**Comment by Masao87** [comment]
Except for the missing Homunculus stuff it all seems to work fine for me at least by now, I guess this can be merged into the master branch? <3

File: src/map/skill.cpp [hunk 6]
```diff
@@ -1150,11 +1148,6 @@ struct s_skill_unit_layout *skill_get_unit_layout(uint16 skill_id, uint16 skill_
 
 struct s_skill_nounit_layout* skill_get_nounit_layout(uint16 skill_id, uint16 skill_lv, struct block_list* src, int x, int y, int dir)
 {
-	if( skill_id == LG_OVERBRAND )
-		return &skill_nounit_layout[overbrand_nounit_pos + dir];
-	else if( skill_id == LG_OVERBRAND_BRANDISH )
-		return &skill_nounit_layout[overbrand_brandish_nounit_pos + dir];
-
 	ShowError("skill_get_nounit_layout: unknown no-unit layout for skill %d (level %d)\n", skill_id, skill_lv);
 	return &skill_nounit_layout[0];
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:1152]: Is it just github again or is this whole function useless now?

File: src/map/skill.cpp [hunk 7]
```diff
@@ -5186,13 +5145,21 @@ int skill_castend_damage_id (struct block_list* src, struct block_list *bl, uint
...
-			if (skill_id == SJ_PROMINENCEKICK) // Trigger the 2nd hit. (100% fire damage.)
-				skill_attack(skill_get_type(skill_id), src, src, bl, skill_id, skill_lv, tick, sflag|8|SD_ANIMATION);
+			switch (skill_id) {
+				case NPC_VAMPIRE_GIFT:
+					if (heal > 0) {
+						clif_skill_nodamage(NULL, src, AL_HEAL, heal, 1);
+						status_heal(src, heal, 0, 0);
+					}
+					break;
+				case SN_SHARPSHOOTING:
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:5156]: ifdef renewal?

File: src/map/skill.cpp [hunk 8]
```diff
@@ -17046,10 +16829,25 @@ struct s_skill_condition skill_get_requirement(struct map_session_data* sd, uint
 							if (i != skill_lv%11 - 1)
 								continue;
 							break;
+#ifdef RENEWAL
 						case AM_CALLHOMUN:
-							if (sd->status.hom_id) //Don't delete items when hom is already out.
```
Comment (Lemongrass3110) [inline src/map/skill.cpp:17050]: What about prerenewal?

**Review by Lemongrass3110 — APPROVED** [review]

**Comment by clashlucifer** [comment]
สกรีนช็อต 2021-05-10 112857

After I updated this bundle, I got some possible bugs with Sorcerer's skill. My friend also tested it, used Earth Summon skills, and used various skills, causing the server to crash instantly. Hope you find this information useful.

**Comment by aleos89** [comment]
This isn't related to this PR. You can find the issue at #5933.

**Pull request merged**